### DISCUSSION TITLE
Difference between select DD1,DD2 and select columns('^DD') and coalesce ?

### DISCUSSION THREAD

**Discussion by fundef1** [duckdb/duckdb#8968] [Q&A] [answered]
Hi,
I have this little test case: 
```
create table test(DD1 varchar,MM1 varchar,DD2 varchar,MM2 varchar);
insert into test values ('01','',NULL,''),(NULL,'','02','');
```
I only want a single DD column, but coalesce in combination with columns returns two columns iso one:
```
select coalesce(columns('^DD')) from test;
┌────────────────────┬────────────────────┐
│ COALESCE(test.DD1) │ COALESCE(test.DD2) │
│      varchar       │      varchar       │
├────────────────────┼────────────────────┤
│ 01                 │                    │
│                    │ 02                 │
└────────────────────┴────────────────────┘
```
while specifying the columns explicitly does give me the correct result.
```
select coalesce(DD1,DD2) from test;
┌────────────────────┐
│ COALESCE(DD1, DD2) │
│      varchar       │
├────────────────────┤
│ 01                 │
│ 02                 │
└────────────────────┘
```

What am I missing/doing wrong? 
I thought the columns() syntax was just an easier sql way to select columns without functional differences:
```
-- select all columns matching the given regex from the table
SELECT COLUMNS('number\d+') FROM addresses;
```

thanks!

**Comment by l1t1** [comment]
see http://duckdb.org/docs/sql/expressions/star#columns
The COLUMNS expression can be used to execute the same expression on multiple columns.

**Comment by fundef1** [comment]
I'm sorry but I don't understand, how that relates to coalesce behaving as I've described.
How would I use coalesce with columns() so as to have it coalesce into a single columns?

**Comment by l1t1** [comment]
for your example. 
```
D from(SELECT COLUMNS('(DD)\d+') FROM test)select coalesce(#1,#2);
┌──────────────────┐
│ COALESCE(#1, #2) │
├──────────────────┤
│ 01               │
│ 02               │
└──────────────────┘
```

**Reply by fundef1** [reply]
That worked! Thanks for the quick response. 
And I learned that you can select columns by index/position with #1, #2 - thanks again!

But now I need to specify the columns in advance, which I'm trying to avoid by using the columns() syntax. 
so, yes it worked but not as I intended... how to use this trick without knowing the (number of) DD columns upfront?

**Comment by fundef1** [comment]
well, I got... something. Far from pretty and I'm confident sure "I'm doing this wrong" (and that it'll break the moment I hit it with some real data). anyway, here is my workaround, written out very verbosely, feedback welcome.

````
DROP TABLE test;

CREATE TABLE test (DD1 VARCHAR, MM1 VARCHAR, DD2 VARCHAR, MM2 VARCHAR);
INSERT INTO test VALUES ('01', '02', '', ''), ('', '', '03', '04');

-- following cases should not happen:
-- re2 stops after a succesfull match
INSERT INTO test VALUES ('05', '06', '07', '08');
-- no match at all is possible, but ok to ignore
INSERT INTO test VALUES ('', '', '', '');

WITH DDs
AS (
	SELECT COLUMNS('^DD')
	FROM test
	), DDu
AS (
	unpivot DDs ON *
	INTO NAME dd VALUE val
	), DD
AS (
	SELECT val AS DD
	FROM DDu
	WHERE strlen(VAL) > 0
	), MMs
AS (
	SELECT COLUMNS('^MM')
	FROM test
	), MMu
AS (
	unpivot MMs ON *
	INTO NAME mm VALUE val
	), MM
AS (
	SELECT val AS MM
	FROM MMu
	WHERE strlen(VAL) > 0
	)
SELECT * FROM DD positional JOIN MM;	
```
results in
```
┌─────────┬─────────┐
│   DD    │   MM    │
│ varchar │ varchar │
├─────────┼─────────┤
│ 01      │ 02      │
│ 03      │ 04      │
│ 05      │ 06      │
│ 07      │ 08      │
└─────────┴─────────┘
```


note: replaced the NULLs with empty '' as that is what the regexp_extract is returning.

**Comment by l1t1** [comment]
your example hasn't a order by, what's the position of result?

**Reply by fundef1** [reply]
I'm not sure I understand your question. 
(I'm assuming) the position remains the same throughout, i.e. there is indeed no (re-)ordering - as that would break things between DD/MM columns as the join is positional.

**Comment by l1t1** [comment, marked as answer]
pivot(select substr(xx,1,2) x,row_number()over(partition by x) rn,val from(unpivot test ON * into name xx VALUE val)where strlen(val)>0) on x using max(val)group by rn;

**Reply by fundef1** [reply]
Thanks for providing a solution without the need to know the number DD columns up front. 

But to be honest, I'll need a bit of time to understand what is really going on here. ;-)
